File: samples/FsBulletML2.Sample.Unity2D.CSharp/Assets/Scripts/Player.cs
using UnityEngine;
using System;
using FsBulletML2;
using R3;
using R3.Triggers;

public class Player : MonoBehaviour
{
    public float speed = 5;
    public GameObject Bullet;
    public GameObject BombType;
    public bool isBomb = true;

    public readonly ReactiveProperty<int> DamageRp = new(0);
    public readonly ReactiveProperty<Vector2> PositionRp = new(Vector2.zero);
    public int Damage => DamageRp.Value;

    // 旧は BulletmlTask（弾幕と実行状態が 1 つ の型）を 3 本 持ち回っていた。
    // 新 API では読み込んだ弾幕（BulletmlScript）だけを持ち、実行位置は
    // 撃つたびに Runner.newRoot で作る —— **1 本 の弾幕から何発でも撃てる。**
    // 旧は同じ task を撃つ弾ぜんぶで共有していて、状態が混ざる形だった
    private static BulletmlScript b2wayLeftBulletScript;
    private static BulletmlScript b2wayRightBulletScript;
    private static BulletmlScript homingScript;

    public float X
    {
        get => PositionRp.Value.x;
        set
        {
            var p = PositionRp.Value;
            p.x = value;
            PositionRp.Value = p;
        }
    }

    public float Y
    {
        get => PositionRp.Value.y;
        set
        {
            var p = PositionRp.Value;
            p.y = value;
            PositionRp.Value = p;
        }
    }

    void Awake()
    {
        // **Init が先。** 読む段の Env（FrontEnv.Load）は BulletMLManager から
        // rand と rank を引くので、口を差し込む前に読むと NullReference になる
        FsBulletML2.BulletMLManager.Init(new BulletFunctions());
        // Bullets の PlayerBullet は Bulletml（DTD の木）を直に持っている。
        // Enemy 側は BulletmlInfo（名前つき）なので .Script(env) を呼ぶが、
        // **どちらも Runner.Load を通る**（BulletmlInfo.Script はその包み）
        b2wayLeftBulletScript = Runner.Load(CSharpWorld.RandFunc, CSharpWorld.LoadRank, FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.b2wayLeftBullet);
        b2wayRightBulletScript = Runner.Load(CSharpWorld.RandFunc, CSharpWorld.LoadRank, FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.b2wayRightBullet);
        homingScript = Runner.Load(CSharpWorld.RandFunc, CSharpWorld.LoadRank, FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.homing);
    }

    void Start()
    {
        PositionRp.Value = new Vector2(transform.position.x, transform.position.y);

        PositionRp
            .Subscribe(p =>
            {
                var pos = transform.position;
                pos.x = p.x;
                pos.y = p.y;
                transform.position = pos;
            })
            .AddTo(this);

        DamageRp
            .Skip(1)
            .Subscribe(_ =>
            {
                if (isBomb) Bomb.GenerateBomb(BombType, transform.position);
            })
            .AddTo(this);

        var update = Observable.EveryUpdate(destroyCancellationToken);

        update
            .Select(_ => (x: Input.GetAxisRaw("Horizontal"), y: Input.GetAxisRaw("Vertical")))
            .Where(v => v.x != 0f || v.y != 0f)
            .Subscribe(v => ApplyMove(v.x, v.y));

        update
            .Where(_ => Input.GetKey(KeyCode.Z))
            .Subscribe(_ =>
            {
                Shoot2WayLeftBullet();
                Shoot2WayRightBullet();
            });

        this.OnTriggerEnter2DAsObservable()
            .Subscribe(_ => HitByEnemyBullet())
            .AddTo(this);
    }

    void ApplyMove(float x, float y)
    {
        var p = PositionRp.Value;
        var mx = p.x + x / 100f * speed;
        if (mx >= 0.4f && mx <= 4.4f)
        {
            p.x = mx;
        }
        var my = p.y + y / 100f * speed;
        if (my > -6.0f && my <= -0.4f)
        {
            p.y = my;
        }
        PositionRp.Value = p;
    }

    private void Shoot2WayLeftBullet()
    {
        var position = this.transform.position + new Vector3(-0.1f, 0.1f, 0);
        BulletEntityFactory.SpawnPlayer(position, Player.b2wayLeftBulletScript);
    }

    private void Shoot2WayRightBullet()
    {
        var position = this.transform.position + new Vector3(0.1f, 0.1f, 0);
        BulletEntityFactory.SpawnPlayer(position, Player.b2wayRightBulletScript);
    }

    private void ShootHomingBullet()
    {
        BulletEntityFactory.SpawnPlayer(this.transform.position, Player.homingScript);
    }

    public void HitByEnemyBullet()
    {
        DamageRp.Value += 1;
    }

    void OnDestroy()
    {
        DamageRp.Dispose();
        PositionRp.Dispose();
    }
}

File: samples/FsBulletML2.Sample.Unity2D.FSharp/FsBulletML2.Sample.Unity2D.FSharp/Player.fs
namespace FsBulletML2.Sample.Unity2D.FSharp

open R3
// R3 のあとに開くこと（`Observable` が両方にある。FrameTicker の但し書き）
open FSharp.Control.R3
open UnityEngine
open FsBulletML2
open FsBulletML2.Unity2D

type Player () =
  inherit MonoBehaviour ()

  /// 受けたダメージ。増えたら爆風（購読は `Start`）。
  /// F# は let 束縛を val より前に置くので、ここに居る
  let damageRp = new ReactiveProperty<int>(0)

  [<DefaultValue>]val mutable public bulletObject : GameObject
  [<DefaultValue>]val mutable public bombType : GameObject
  [<DefaultValue>]val mutable public speed : float32
  [<DefaultValue>]val mutable public isBomb : bool

  [<DefaultValue>]val mutable private b2wayLeftBulletTask : BulletmlScript option
  [<DefaultValue>]val mutable private b2wayRightBulletTask : BulletmlScript option
  [<DefaultValue>]val mutable private hommingTask : BulletmlScript option

  interface IPlayerPosition with
    member this.PlayerPosX () = this.transform.position.x
    member this.PlayerPosY () = this.transform.position.y

  /// 受けたダメージ。Informations が読む
  member this.Damage = damageRp.Value
  member this.DamageRp = damageRp

  member this.Awake () =
    // Init が先。 読む段の Env は BulletMLManager から rand と rank を
    // 引くので、口を差し込む前に読むと NullReference になる
    BulletMLManager.Init(new BulletFunctions(this))
    this.b2wayLeftBulletTask <- Runner.load loadRand (loadRank ()) FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.b2wayLeftBullet |> Some
    this.b2wayRightBulletTask <- Runner.load loadRand (loadRank ()) FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.b2wayRightBullet |> Some
    this.hommingTask <- Runner.load loadRand (loadRank ()) FsBulletML2.Bullets.Dsl.PlayerBullet.PlayerBullet.homing |> Some

  member this.X with get () = this.transform.position.x
                 and set (v) = this.transform.position <- Vector3(v, this.transform.position.y, this.transform.position.z)
  member this.Y with get () = this.transform.position.y
                 and set (v) = this.transform.position <- Vector3(this.transform.position.x, v, this.transform.position.z)

  /// 毎コマ の仕事を 4 本 の流れに割る。
  member this.Start () =
    let update = FrameTicker.Frames

    // 1. 移動。入力が入っているコマだけ通す
    update
    |> Observable.map (fun _ -> struct (Input.GetAxisRaw "Horizontal", Input.GetAxisRaw "Vertical"))
    |> Observable.filter (fun struct (x, y) -> x <> 0.0f || y <> 0.0f)
    |> subscribeUntilDestroy this (fun struct (x, y) -> this.ApplyMove x y)

    // 2. Z を押している間ずっと 2way
    update
    |> Observable.filter (fun _ -> Input.GetKey KeyCode.Z)
    |> subscribeUntilDestroy this (fun _ ->
        this.Shoot2WayLeftBullet ()
        this.Shoot2WayRightBullet ())

    // 3. ホーミングは 61 コマ に 1 回 だけ。
    update
    |> Observable.mapi (fun i _ -> i)
    |> Observable.filter (fun i -> i % 61 = 60 && Input.GetKey KeyCode.Z)
    |> subscribeUntilDestroy this (fun _ -> this.ShootHomingBullet ())

    // 4. ダメージが増えたら爆風。`skip 1` は初期値の 0 を捨てるため
    //    （ReactiveProperty は購読した瞬間に現在値を 1 個 流す）
    (damageRp :> Observable<int>)
    |> Observable.skip 1
    |> subscribeUntilDestroy this (fun _ ->
        if this.isBomb then Bomb.GenerateBomb(this.bombType, this.transform.position))

  member private this.ApplyMove (x: float32) (y: float32) =
    let mx = this.X + x / 100.f * this.speed
    if (mx >= 0.4f && mx <= 4.4f) then
        this.X <- mx

    let my = this.Y + y / 100.f * this.speed
    if (my > -6.0f && my <= -0.4f) then
        this.Y <- my

  /// 弾を 1 発 撃つ。prefab ではなく Entity を作る。
  /// 台本が無ければ何もしない（Awake が走る前に呼ばれた場合）
  member private this.Fire (position: Vector3) (script: BulletmlScript option) =
    match script with
    | Some s -> BulletEntityFactory.SpawnPlayer(position, s) |> ignore
    | None -> ()

  member this.Shoot2WayLeftBullet () =
    this.Fire (this.transform.position + new Vector3(-0.1f, 0.1f, 0.f)) this.b2wayLeftBulletTask

  member this.Shoot2WayRightBullet () =
    this.Fire (this.transform.position + new Vector3(0.1f, 0.1f, 0.f)) this.b2wayRightBulletTask

  /// 間隔の判定はここに無い。 持っているのは流れの側（`Start` の 3 番）
  member this.ShootHomingBullet () =
    this.Fire this.transform.position this.hommingTask

  /// 敵弾が当たった。
  /// 当たり判定は BulletEcsDriver がやる —— ECS の弾は Collider2D を持たないので、OnTriggerEnter2D は届かない。
  member this.HitByEnemyBullet () =
    damageRp.Value <- damageRp.Value + 1

  member this.OnTriggerEnter2D (collier:Collider2D) =
    // GameObject の弾（もう出ないが、prefab が残っている経路）向け
    this.HitByEnemyBullet()

  member this.OnDestroy () = damageRp.Dispose()
